src/pycsou/opt/solver/cg.py

```diff
@@ -72,10 +72,18 @@ def fit(
            * MANUAL: fit() + steps()
         """
         if stop_crit is None:
+
+            def explicit_residual(primal):
```

Comment:
You can get away without having to call `get_array_module()` each per iteration.

```
def explicit_residual(primal):
    mst = self._mstate  # shorthand
    residual = mst["b"].copy()
    residual -= self._A.apply(primal)
    return residual
```